File: sdks/go/plexspaces/agent.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// AgentLoop standalone utility — OODA-loop agent harness for PlexSpaces Go SDK.
//
// Provides Go parity with Python/Rust/TypeScript AgentLoop:
// - Durable step recording (each Observe/Orient/Decide/Act journaled in AgentTrajectory)
// - Token budget enforcement (BudgetExceeded() halts loop when cumulative tokens >= budget)
// - Iteration limit enforcement (IterationLimitReached() halts loop after N iterations)
// - Human-in-the-loop suspend (Suspend sets IsSuspended to true)
// - Trajectory capture (FinalizeTrajectory / GetTrajectory)
//
// AgentLoop is a standalone plain struct — it is NOT a decorator and does NOT wrap
// other types. Embed it in an actor struct or hold it as a field.
//
// Usage:
//
//	type ResearchAgent struct {
//	    plexspaces.BaseActor
//	    loop *plexspaces.AgentLoop
//	}
//
//	func (a *ResearchAgent) Handle(from, msgType, payloadJSON string) string {
//	    cfg := plexspaces.DefaultAgentConfig()
//	    cfg.MaxIterations = 5
//	    a.loop = plexspaces.NewAgentLoop("research-agent-01", cfg)
//	    for !a.loop.IterationLimitReached() && !a.loop.BudgetExceeded() {
//	        obs := a.loop.Observe(payloadJSON)
//	        plan := a.loop.Orient(obs)
//	        action := a.loop.Decide(plan)
//	        result := a.loop.Act(action, 100, 50, "gpt-4")
//	        a.loop.IncrementIteration()
//	        _ = result
//	        break
//	    }
//	    traj := a.loop.FinalizeTrajectory("success", "")
//	    data, _ := json.Marshal(traj)
//	    return string(data)
//	}

package plexspaces

import (
	"fmt"
	"time"
)

// ========================================================================
// AgentStepKind constants
// ========================================================================

// AgentStepKind identifies the OODA phase or special event for a trajectory step.
// Aligns with Python AgentLoop step kinds.
type AgentStepKind string

const (
	// StepKindObserve is the Observe phase: gather information from environment, memory, context.
	StepKindObserve AgentStepKind = "observe"
	// StepKindOrient is the Orient phase: analyse and contextualise observations.
	StepKindOrient AgentStepKind = "orient"
	// StepKindDecide is the Decide phase: select next action from candidate plans.
	StepKindDecide AgentStepKind = "decide"
	// StepKindAct is the Act phase: execute the chosen action (often an LLM call).
	StepKindAct AgentStepKind = "act"
	// StepKindToolCall records a single tool invocation within an Act phase.
	StepKindToolCall AgentStepKind = "tool_call"
	// StepKindSuspend records agent suspension for human-in-the-loop or external signal.
	StepKindSuspend AgentStepKind = "suspend"
)

// ========================================================================
// AgentStep
// ========================================================================

// AgentStep records a single OODA step or event within an AgentTrajectory.
// Aligns with Python AgentStep and proto trajectory step shape.
type AgentStep struct {
	// StepID is a unique ULID-like identifier for this step.
	StepID string `json:"step_id"`
	// Kind is the OODA phase or event kind (observe, orient, decide, act, …).
	Kind AgentStepKind `json:"kind"`
	// Method is the handler method name that produced this step.
	Method string `json:"method"`
	// Input is the data passed into this step.
	Input any `json:"input,omitempty"`
	// Output is the data returned by this step.
	Output any `json:"output,omitempty"`
	// StartedAtMs is the Unix epoch timestamp in milliseconds when the step started.
	StartedAtMs int64 `json:"started_at_ms"`
	// CompletedAtMs is the Unix epoch timestamp in milliseconds when the step completed.
	CompletedAtMs int64 `json:"completed_at_ms"`
	// DurationMs is the wall-clock duration of the step in milliseconds.
	DurationMs int64 `json:"duration_ms"`
	// Success is true when the step completed without error.
	Success bool `json:"success"`
	// Error holds the error message when Success is false.
	Error string `json:"error,omitempty"`
	// ToolName is the tool identifier for StepKindToolCall steps.
	ToolName string `json:"tool_name,omitempty"`
	// InputTokens is the number of prompt tokens consumed in this step.
	InputTokens int `json:"input_tokens"`
	// OutputTokens is the number of completion tokens produced in this step.
	OutputTokens int `json:"output_tokens"`
	// Model is the LLM model identifier used in this step (empty for non-LLM steps).
	Model string `json:"model,omitempty"`
	// Metadata holds arbitrary string key-value annotations for this step.
	Metadata map[string]string `json:"metadata,omitempty"`
}

// ========================================================================
// AgentTrajectory
// ========================================================================

// AgentTrajectory is the complete execution record for one agent run.
// Aligns with Python AgentTrajectory and the proto trajectory message shape.
type AgentTrajectory struct {
	// TrajectoryID is a unique ULID-like identifier for this trajectory.
	TrajectoryID string `json:"trajectory_id"`
	// AgentActorID is the actor identifier of the agent that produced this trajectory.
	AgentActorID string `json:"agent_actor_id"`
	// EvalRunID links this trajectory to an evaluation run (empty in production).
	EvalRunID string `json:"eval_run_id,omitempty"`
	// ScenarioID links this trajectory to an evaluation scenario (empty in production).
	ScenarioID string `json:"scenario_id,omitempty"`
	// Steps is the ordered list of recorded OODA steps.
	Steps []AgentStep `json:"steps"`
	// Outcome is the final outcome label: "running", "success", "failure", "suspended", etc.
	Outcome string `json:"outcome"`
	// OutcomeDetail provides human-readable context for the outcome.
	OutcomeDetail string `json:"outcome_detail,omitempty"`
	// TotalInputTokens is the cumulative prompt tokens across all steps.
	TotalInputTokens int `json:"total_input_tokens"`
	// TotalOutputTokens is the cumulative completion tokens across all steps.
	TotalOutputTokens int `json:"total_output_tokens"`
	// StartedAtMs is the Unix epoch timestamp in milliseconds when the trajectory started.
	StartedAtMs int64 `json:"started_at_ms"`
	// CompletedAtMs is the Unix epoch timestamp in milliseconds when the trajectory was finalised.
	CompletedAtMs int64 `json:"completed_at_ms"`
	// DurationMs is the total wall-clock duration of the trajectory in milliseconds.
	DurationMs int64 `json:"duration_ms"`
	// Score is an optional numeric quality score assigned by an evaluator.
	Score float64 `json:"score"`
	// Metadata holds arbitrary string key-value annotations for this trajectory.
	Metadata map[string]string `json:"metadata,omitempty"`
}

// ========================================================================
// AgentConfig
// ========================================================================

// AgentConfig controls agent loop behaviour.
// Aligns with the parameters of the Python AgentConfig(max_iterations=…, token_budget=…).
type AgentConfig struct {
	// MaxIterations is the maximum number of OODA loop iterations before the agent stops.
	// Zero is treated as no limit (not recommended for production).
	MaxIterations int
	// TokenBudget is the cumulative token limit across all steps.
	// Zero means unlimited.
	TokenBudget int
	// EvalRunID optionally links trajectory to an evaluation run.
	EvalRunID string
	// ScenarioID optionally links trajectory to an evaluation scenario.
	ScenarioID string
}

// DefaultAgentConfig returns a safe production default: 10 iterations, unlimited tokens.
func DefaultAgentConfig() AgentConfig {
	return AgentConfig{
		MaxIterations: 10,
		TokenBudget:   0,
	}
}

// ========================================================================
// AgentLoop
// ========================================================================

// AgentLoop is a standalone OODA-loop harness.
//
// Embed AgentLoop in an actor struct (or hold it as a field) to gain
// structured step recording, budget/iteration guards, and trajectory export.
// All state is held in the struct — no global state.
//
// Example:
//
//	cfg := plexspaces.DefaultAgentConfig()
//	loop := plexspaces.NewAgentLoop("my-actor-id", cfg)
//	for !loop.IterationLimitReached() && !loop.BudgetExceeded() {
//	    obs   := loop.Observe(input)
//	    plan  := loop.Orient(obs)
//	    action := loop.Decide(plan)
//	    result := loop.Act(action, 120, 60, "claude-3-5-sonnet")
//	    loop.IncrementIteration()
//	    _ = result
//	}
//	traj := loop.FinalizeTrajectory("success", "")
type AgentLoop struct {
	config         AgentConfig
	trajectory     AgentTrajectory
	iterationCount int
	isSuspended    bool
	stepCounter    uint64
}

// NewAgentLoop creates a new AgentLoop with the given actorID and config.
// actorID is stored in trajectory.AgentActorID to link the trajectory to the producing actor.
// The trajectory is initialised with "running" outcome and the current timestamp.
func NewAgentLoop(actorID string, cfg AgentConfig) *AgentLoop {
	return &AgentLoop{
		config: cfg,
		trajectory: AgentTrajectory{
			TrajectoryID: newID(),
			AgentActorID: actorID,
			EvalRunID:    cfg.EvalRunID,
			ScenarioID:   cfg.ScenarioID,
			Outcome:      "running",
			Steps:        []AgentStep{},
			StartedAtMs:  nowMs(),
		},
	}
}

// ========================================================================
// OODA step methods
// ========================================================================

// Observe records a StepKindObserve step and returns input unchanged.
// Use this to journal the raw environment/context data gathered at the start of each OODA cycle.
func (a *AgentLoop) Observe(input any) any {
	a.recordStep(string(StepKindObserve), "observe", input, input, true, "", "", "", 0, 0)
	return input
}

// Orient records a StepKindOrient step and returns obs unchanged.
// Use this to journal the analysis / contextualisation of the observation.
func (a *AgentLoop) Orient(obs any) any {
	a.recordStep(string(StepKindOrient), "orient", obs, obs, true, "", "", "", 0, 0)
	return obs
}

// Decide records a StepKindDecide step and returns plan unchanged.
// Use this to journal the selected action plan before execution.
func (a *AgentLoop) Decide(plan any) any {
	a.recordStep(string(StepKindDecide), "decide", plan, plan, true, "", "", "", 0, 0)
	return plan
}

// Act records a StepKindAct step with token counts and model, and returns action unchanged.
// inputTokens and outputTokens are added to the trajectory cumulative totals.
func (a *AgentLoop) Act(action any, inputTokens, outputTokens int, model string) any {
	a.recordStep(string(StepKindAct), "act", action, action, true, "", "", model, inputTokens, outputTokens)
	return action
}

// ToolCall records a StepKindToolCall step with tool name, arguments, result, tokens, and model.
// Returns result unchanged for ergonomic chaining.
func (a *AgentLoop) ToolCall(toolName string, arguments, result any, inputTokens, outputTokens int, model string) any {
	a.recordStep(string(StepKindToolCall), "tool_call", arguments, result, true, "", toolName, model, inputTokens, outputTokens)
	return result
}

// Suspend records a StepKindSuspend step with the given reason and sets isSuspended to true.
// Subsequent calls to IsSuspended() will return true.
func (a *AgentLoop) Suspend(reason string) {
	a.isSuspended = true
	a.recordStep(string(StepKindSuspend), "suspend", reason, nil, true, "", "", "", 0, 0)
}

// IsSuspended reports whether Suspend has been called on this loop.
func (a *AgentLoop) IsSuspended() bool {
	return a.isSuspended
}

// BudgetExceeded reports whether cumulative tokens have reached or exceeded the configured
// TokenBudget. Returns false when TokenBudget is zero (unlimited).
func (a *AgentLoop) BudgetExceeded() bool {
	if a.config.TokenBudget <= 0 {
		return false
	}
	total := a.trajectory.TotalInputTokens + a.trajectory.TotalOutputTokens
	return total >= a.config.TokenBudget
}

// IterationLimitReached reports whether the iteration count has reached or exceeded
// MaxIterations. Returns false when MaxIterations is zero (unlimited).
func (a *AgentLoop) IterationLimitReached() bool {
	if a.config.MaxIterations <= 0 {
		return false
	}
	return a.iterationCount >= a.config.MaxIterations
}

// IncrementIteration advances the iteration counter by one.
// Call this once per OODA loop cycle, typically at the end of the loop body.
func (a *AgentLoop) IncrementIteration() {
	a.iterationCount++
}

// FinalizeTrajectory closes the trajectory with the given outcome and detail,
// sets CompletedAtMs and DurationMs, and returns a snapshot copy.
// Subsequent calls to GetTrajectory will reflect the finalised state.
func (a *AgentLoop) FinalizeTrajectory(outcome, detail string) AgentTrajectory {
	a.trajectory.Outcome = outcome
	a.trajectory.OutcomeDetail = detail
	a.trajectory.CompletedAtMs = nowMs()
	a.trajectory.DurationMs = a.trajectory.CompletedAtMs - a.trajectory.StartedAtMs
	return a.trajectory
}

// GetTrajectory returns a snapshot of the trajectory in its current state.
// Safe to call before FinalizeTrajectory for a mid-run inspection.
func (a *AgentLoop) GetTrajectory() AgentTrajectory {
	return a.trajectory
}

// ========================================================================
// Private helpers
// ========================================================================

// recordStep creates and appends an AgentStep to the trajectory.
// It updates cumulative token totals and returns the recorded step.
func (a *AgentLoop) recordStep(
	kind, method string,
	input, output any,
	success bool,
	errorStr, toolName, model string,
	inputTokens, outputTokens int,
) AgentStep {
	start := nowMs()
	end := nowMs()
	step := AgentStep{
		StepID:        a.nextStepID(),
		Kind:          AgentStepKind(kind),
		Method:        method,
		Input:         input,
		Output:        output,
		StartedAtMs:   start,
		CompletedAtMs: end,
		DurationMs:    end - start,
		Success:       success,
		Error:         errorStr,
		ToolName:      toolName,
		Model:         model,
		InputTokens:   inputTokens,
		OutputTokens:  outputTokens,
	}
	a.trajectory.Steps = append(a.trajectory.Steps, step)
	a.trajectory.TotalInputTokens += inputTokens
	a.trajectory.TotalOutputTokens += outputTokens
	return step
}

// nowMs returns the current Unix epoch time in milliseconds.
func nowMs() int64 {
	return time.Now().UnixMilli()
}

// newID generates a unique ID for a step or trajectory.
// Uses a combination of the current nanosecond timestamp and a per-instance
// counter to guarantee uniqueness without global state.
func newID() string {
	return fmt.Sprintf("%016x", time.Now().UnixNano())
}

// nextStepID increments the per-instance counter and returns a unique ID.
func (a *AgentLoop) nextStepID() string {
	a.stepCounter++
	return fmt.Sprintf("%016x%08x", time.Now().UnixNano(), a.stepCounter)
}

File: sdks/python/plexspaces/agent.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (C) 2025 PlexSpaces Contributors
#
# This file is part of PlexSpaces.
#
# PlexSpaces is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option)
# any later version.
#
# PlexSpaces is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
# details.
#
# You should have received a copy of the GNU Affero General Public License
# along with PlexSpaces. If not, see <https://www.gnu.org/licenses/>.

"""
Agent harness utilities for PlexSpaces Python SDK.

``AgentLoop`` is a **standalone class** (no decorator magic, no class injection)
that provides structured OODA step tracking, token budget enforcement, and
suspend/resume helpers.  Use it with ``@workflow_actor`` for durable agent
execution.

Example::

    from plexspaces import workflow_actor, state, run_handler, host
    from plexspaces.agent import AgentLoop, AgentConfig

    @workflow_actor
    class ResearchAgent:
        task: str = state(default="")

        @run_handler
        def run(self, task: str = "") -> dict:
            loop = AgentLoop(
                actor_id=host.actor_id(),
                config=AgentConfig(max_iterations=10, token_budget=4096),
            )
            while not loop.iteration_limit_reached() and not loop.budget_exceeded():
                obs    = loop.observe({"task": task})
                plan   = loop.orient(obs)
                action = loop.decide(plan)
                result = loop.act(action)
                loop.increment_iteration()
                if result.get("done"):
                    break
            return loop.finalize_trajectory("completed", "task done")
"""

from __future__ import annotations

import time
import uuid
from dataclasses import dataclass
from typing import Any, Dict, List, Optional


# ---------------------------------------------------------------------------
# Utilities
# ---------------------------------------------------------------------------

def _now_ms() -> int:
    """Current time as integer milliseconds since the Unix epoch."""
    return int(time.time() * 1000)


def _new_ulid() -> str:
    """Generate a new ULID-compatible unique string (UUID4 fallback)."""
    return str(uuid.uuid4())


# ---------------------------------------------------------------------------
# AgentStepKind constants
# ---------------------------------------------------------------------------

class AgentStepKind:
    OBSERVE = "observe"
    ORIENT = "orient"
    DECIDE = "decide"
    ACT = "act"
    TOOL_CALL = "tool_call"
    SUSPEND = "suspend"


# ---------------------------------------------------------------------------
# AgentStep
# ---------------------------------------------------------------------------

class AgentStep:
    """A single step in an agent's OODA trajectory."""

    def __init__(self, kind: str, method: str, input_data: Any = None):
        self.step_id: str = _new_ulid()
        self.kind: str = kind
        self.method: str = method
        self.input: Any = input_data
        self.output: Any = None
        self.started_at_ms: int = _now_ms()
        self.completed_at_ms: Optional[int] = None
        self.success: bool = False
        self.error: Optional[str] = None
        self.input_tokens: int = 0
        self.output_tokens: int = 0
        self.model: str = ""
        self.metadata: Dict[str, str] = {}

    def complete(
        self,
        output: Any,
        input_tokens: int = 0,
        output_tokens: int = 0,
        model: str = "",
    ) -> "AgentStep":
        """Mark the step as successfully completed."""
        self.output = output
        self.completed_at_ms = _now_ms()
        self.success = True
        self.input_tokens = input_tokens
        self.output_tokens = output_tokens
        self.model = model
        return self

    def fail(self, error: str) -> "AgentStep":
        """Mark the step as failed."""
        self.completed_at_ms = _now_ms()
        self.success = False
        self.error = error
        return self

    @property
    def duration_ms(self) -> int:
        end = self.completed_at_ms or _now_ms()
        return max(0, end - self.started_at_ms)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "step_id": self.step_id,
            "kind": self.kind,
            "method": self.method,
            "input": self.input,
            "output": self.output,
            "started_at_ms": self.started_at_ms,
            "completed_at_ms": self.completed_at_ms,
            "duration_ms": self.duration_ms,
            "success": self.success,
            "error": self.error,
            "input_tokens": self.input_tokens,
            "output_tokens": self.output_tokens,
            "model": self.model,
            "metadata": self.metadata,
        }


# ---------------------------------------------------------------------------
# AgentTrajectory
# ---------------------------------------------------------------------------

class AgentTrajectory:
    """Complete trajectory for one agent run (all OODA steps, token totals)."""

    def __init__(
        self,
        agent_actor_id: str,
        eval_run_id: str = "",
        scenario_id: str = "",
    ):
        self.trajectory_id: str = _new_ulid()
        self.agent_actor_id: str = agent_actor_id
        self.eval_run_id: str = eval_run_id
        self.scenario_id: str = scenario_id
        self.steps: List[AgentStep] = []
        self.outcome: str = ""
        self.outcome_detail: str = ""
        self.started_at_ms: int = _now_ms()
        self.completed_at_ms: Optional[int] = None
        self.total_input_tokens: int = 0
        self.total_output_tokens: int = 0

    @property
    def duration_ms(self) -> int:
        end = self.completed_at_ms or _now_ms()
        return max(0, end - self.started_at_ms)

    @property
    def step_count(self) -> int:
        return len(self.steps)

    def add_step(self, step: AgentStep) -> None:
        """Append a completed step and accumulate token totals."""
        self.steps.append(step)
        self.total_input_tokens += step.input_tokens
        self.total_output_tokens += step.output_tokens

    def complete(self, outcome: str, detail: str = "") -> None:
        """Finalise the trajectory with an outcome string."""
        self.outcome = outcome
        self.outcome_detail = detail
        self.completed_at_ms = _now_ms()

    def to_dict(self) -> Dict[str, Any]:
        return {
            "trajectory_id": self.trajectory_id,
            "agent_actor_id": self.agent_actor_id,
            "eval_run_id": self.eval_run_id,
            "scenario_id": self.scenario_id,
            "steps": [s.to_dict() for s in self.steps],
            "step_count": len(self.steps),
            "outcome": self.outcome,
            "outcome_detail": self.outcome_detail,
            "started_at_ms": self.started_at_ms,
            "completed_at_ms": self.completed_at_ms,
            "duration_ms": self.duration_ms,
            "total_input_tokens": self.total_input_tokens,
            "total_output_tokens": self.total_output_tokens,
        }


# ---------------------------------------------------------------------------
# AgentConfig
# ---------------------------------------------------------------------------

@dataclass
class AgentConfig:
    """Configuration for an AgentLoop."""
    max_iterations: int = 10
    token_budget: int = 0          # 0 = no budget limit


# ---------------------------------------------------------------------------
# AgentLoop
# ---------------------------------------------------------------------------

class AgentLoop:
    """
    Standalone agent harness utility.

    Records OODA steps, enforces token budget and iteration limits, and
    produces an ``AgentTrajectory`` on finalisation.

    This is a **plain class** — there is no class injection or metaclass magic.
    Use it inside a ``@workflow_actor`` ``run_handler``.

    :param actor_id: ID of the hosting actor.
    :param config:   Optional ``AgentConfig`` (max_iterations, token_budget).
    :param eval_run_id: Optional eval run ID to tag the trajectory.
    :param scenario_id: Optional scenario ID to tag the trajectory.
    """

    def __init__(
        self,
        actor_id: str,
        config: Optional[AgentConfig] = None,
        eval_run_id: str = "",
        scenario_id: str = "",
    ):
        self._actor_id = actor_id
        self._config = config or AgentConfig()
        self._trajectory = AgentTrajectory(
            agent_actor_id=actor_id,
            eval_run_id=eval_run_id,
            scenario_id=scenario_id,
        )
        self._iteration_count: int = 0
        self.is_suspended: bool = False

    # ── Private helpers ──────────────────────────────────────────────────────

    def _record(
        self,
        kind: str,
        input_data: Any,
        output: Any = None,
        input_tokens: int = 0,
        output_tokens: int = 0,
        model: str = "",
        error: Optional[str] = None,
    ) -> AgentStep:
        started = _now_ms()
        step = AgentStep(kind=kind, method=kind, input_data=input_data)
        step.started_at_ms = started
        if error:
            step.fail(error)
        else:
            step.complete(
                output=output,
                input_tokens=input_tokens,
                output_tokens=output_tokens,
                model=model,
            )
        self._trajectory.add_step(step)
        return step

    # ── OODA step methods ─────────────────────────────────────────────────────

    def observe(self, input_data: Any, **kwargs: Any) -> Any:
        """Record an OBSERVE step.  Returns *input_data* unchanged."""
        self._record(AgentStepKind.OBSERVE, input_data, output=input_data, **kwargs)
        return input_data

    def orient(self, input_data: Any, **kwargs: Any) -> Any:
        """Record an ORIENT step.  Returns *input_data* unchanged."""
        self._record(AgentStepKind.ORIENT, input_data, output=input_data, **kwargs)
        return input_data

    def decide(self, input_data: Any, **kwargs: Any) -> Any:
        """Record a DECIDE step.  Returns *input_data* unchanged."""
        self._record(AgentStepKind.DECIDE, input_data, output=input_data, **kwargs)
        return input_data

    def act(
        self,
        input_data: Any,
        output: Any = None,
        input_tokens: int = 0,
        output_tokens: int = 0,
        model: str = "",
        **kwargs: Any,
    ) -> Any:
        """Record an ACT step.  Returns *output* (or *input_data* if output is None)."""
        result = output if output is not None else input_data
        self._record(
            AgentStepKind.ACT,
            input_data,
            output=result,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            model=model,
        )
        return result

    def tool_call(
        self,
        name: str,
        args: Any,
        result: Any,
        started_at_ms: Optional[int] = None,
        input_tokens: int = 0,
        output_tokens: int = 0,
        model: str = "",
    ) -> Any:
        """Record a TOOL_CALL step.  Returns *result*."""
        step = AgentStep(
            kind=AgentStepKind.TOOL_CALL,
            method=name,
            input_data=args,
        )
        if started_at_ms is not None:
            step.started_at_ms = started_at_ms
        step.complete(
            output=result,
            input_tokens=input_tokens,
            output_tokens=output_tokens,
            model=model,
        )
        self._trajectory.add_step(step)
        return result

    # ── Control ───────────────────────────────────────────────────────────────

    def suspend(self, reason: str) -> None:
        """Mark the agent as suspended (waiting for external signal)."""
        self.is_suspended = True
        self._record(AgentStepKind.SUSPEND, reason)

    def increment_iteration(self) -> None:
        """Increment the OODA cycle counter.  Call once per full OODA cycle."""
        self._iteration_count += 1

    def budget_exceeded(self) -> bool:
        """True if the cumulative token count meets or exceeds the budget."""
        budget = self._config.token_budget
        if budget <= 0:
            return False
        total = (
            self._trajectory.total_input_tokens
            + self._trajectory.total_output_tokens
        )
        return total >= budget

    def iteration_limit_reached(self) -> bool:
        """True if the OODA cycle count meets or exceeds max_iterations.

        Returns False when max_iterations is 0 (unlimited), matching Go/TS/Rust.
        """
        if self._config.max_iterations <= 0:
            return False
        return self._iteration_count >= self._config.max_iterations

    # ── Trajectory accessors ──────────────────────────────────────────────────

    def get_trajectory(self) -> AgentTrajectory:
        """Return the in-progress ``AgentTrajectory`` object."""
        return self._trajectory

    def finalize_trajectory(self, outcome: str, detail: str = "") -> Dict[str, Any]:
        """
        Finalise the trajectory with an outcome and return it as a dict.

        :param outcome: "completed", "error", "timeout", "budget_exceeded", etc.
        :param detail:  Human-readable detail.
        :returns: ``AgentTrajectory.to_dict()``
        """
        self._trajectory.complete(outcome=outcome, detail=detail)
        return self._trajectory.to_dict()
